lib/pbench/server/api/resources/datasets_metadata.py

```diff
@@ -149,7 +150,22 @@ def _put(self, params: ApiParams, _) -> Response:
                     role = API_OPERATION.UPDATE
         self._check_authorization(str(dataset.owner_id), dataset.access, role)
 
+        # Validate the metadata key values in a separate pass so that we can
+        # fail before committing any changes to the database.
         failures = []
+        for k, v in metadata.items():
+            try:
+                Metadata.validate(dataset=dataset, key=k, value=v)
+            except MetadataBadValue as e:
+                failures.append(str(e))
```

Comment:
Should we be decorating the exception message with the key value (and, perhaps, the corresponding value value) before appending it to the list?  (That is, if we get failures on only a subset of the keys, the requestor would probably appreciate knowing which ones went bad.)

Also, it might not be worth it in this context, but in general we should consider de-duplicating the failure list in code like this.

Also, also, for the sake of testing, we might want to consider sorting the failures prior to returning them, so that they appear in a deterministic order.

Reply:
Note that `MetadataBadValue` *does* have those things, and this code appends the exception string for each error in the loop.

But, yeah; in theory the `.items()` doesn't have a fixed order, so sorting the list before doing the `join` might be a good idea.